### ISSUE TITLE
[HOLD for payment 2025-01-02] [$250] Search - Search loads infinitely when search query contains type:chat and category query

### ISSUE THREAD

**Issue by IuliiaHerets** [Expensify/App#52907]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.65-1
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**Email or phone of affected tester (no customers):** applausetester+kh0811007@applause.expensifail.com
**Issue reported by:** Applause Internal Team

## Action Performed:
1. Go to staging.new.expensify.com
2. Click search router icon.
3. Enter the following query: type:chat in:<any chat> category:car
4. Hit Enter.

## Expected Result:
Search will not load infinitely.

## Actual Result:
Search loads infinitely.

## Workaround:
Unknown

## Platforms:

- [x] Android: Standalone
- [x] Android: HybridApp
- [x] Android: mWeb Chrome
- [x] iOS: Standalone
- [x] iOS: HybridApp
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

https://github.com/user-attachments/assets/14013d8a-7a42-482d-9a1f-6cee8ab021e1

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021859790661845309370</li>
        <li>Upwork Job ID: 1859790661845309370</li>
        <li>Last Price Increase: 2024-12-06</li>
<li>Automatic offers: </li>
<ul>
<li>daledah | Contributor | 105274097</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @Christinadobrzyn</details>

**Comment by Christinadobrzyn** [comment]
I think this might have to do with how `type:chat in:<any chat> category:car` is being read? Here's what is showing under keyword, I think this is supposed to be 'chat'? 

<img width="1720" alt="Snagit 2021 2024-11-22 10 46 13" src="https://github.com/user-attachments/assets/19d6e9bf-41f9-4246-be3b-cdaa6f0d1b56">

**Comment by Anaslancer** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Search - Search loads infinitely when search query contains type:chat and category query

### What is the root cause of that problem?
The `Search` api returns `jsonCode: 402` and empty `onyxData`.
then
https://github.com/Expensify/App/blob/5f483dc37ccfbb967d8eeb49302d3dfa010f93e9/src/components/Search/index.tsx#L190-L195
`shouldShowLoadingState` is `true` because `searchResults?.search?.status` is `undefined`.
According to this below logic, we are showing the infinite loading component.
https://github.com/Expensify/App/blob/5f483dc37ccfbb967d8eeb49302d3dfa010f93e9/src/components/Search/index.tsx#L274-L281

### What changes do you think we should make in order to solve the problem?
We have to add the `failureData` in [`getOnyxLoadingData`](https://github.com/Expensify/App/blob/5f483dc37ccfbb967d8eeb49302d3dfa010f93e9/src/libs/actions/Search.ts#L26) function.
```
    const failureData: OnyxUpdate[] = [
        {
            onyxMethod: Onyx.METHOD.MERGE,
            key: `${ONYXKEYS.COLLECTION.SNAPSHOT}${hash}`,
            value: {
                search: {
                    ...(status && { status }),
                },
                data: undefined,
            },
        }
    ];
```

### What alternative solutions did you explore? (Optional)
Or the backend should send the correct result in `Search` api.

### Contributor details
Your Expensify account email: anasup1995@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01aff093c9a804b145

**Comment by Christinadobrzyn** [comment]
@alitoshmatov can you review these proposals? Thanks!

**Comment by alitoshmatov** [comment]
> that make the condition `shouldShowLoadingState`  is still false

I assume you meant `isDataLoaded` is still false, your RCA is correct. Can you elaborate how you want to show error. I like the idea of adding `failureData` to handle any errors in api.

**Comment by alitoshmatov** [comment]
@Anaslancer Thank you for your proposal, your solution is the same as @daledah 's.

**Comment by alitoshmatov** [comment]
@huult Thank you for detailed proposal, your RCA is correct. But I don't think your solution is future-proof, your solution does solve this issue, but when backend responds with another error the same thing happens again.

**Comment by huult** [comment]
@alitoshmatov Thank you for your feedback.

This issue occurred because the app sent a query with incorrect syntax to the backend, which the backend could not process, resulting in an error.

Therefore, I think we should validate the query to ensure the correct syntax is sent to the backend. With this approach, we can perform the search; if not, we can handle the error by hiding the loading indicator and showing a 'not found' message. However, this might introduce a new issue where the category is sent, but the search result is 'not found'.

> But I don't think your solution is future-proof,

With this concern, I suggest we confirm with the internal team to get information about the filter key that the backend can support. 

For example:

type = expense support key: category, tag...
type = chat support key: from, to...
type = trip support key: from, to...

to we can create the correct query syntax.

@alitoshmatov If you agree with this, we can discuss it in detail and define the next steps

**Comment by alitoshmatov** [comment]
@huult I still think validating query is partial solution, we can't possibly cover all cases where backend sends an error. Moreover this issue is rare since it requires user to type all filters by hand. I also got similar case where even with valid search query backend is responding with an empty data which is causing infinite loading

<img width="1512" alt="Screenshot 2024-12-02 at 4 06 56 PM" src="https://github.com/user-attachments/assets/9f5adcdc-4554-49a7-9b1a-8ad64d901722">
<img width="1512" alt="Screenshot 2024-12-02 at 4 07 10 PM" src="https://github.com/user-attachments/assets/7a287875-2b9a-4673-9edf-8c4171622fae">

**Comment by alitoshmatov** [comment]
@daledah We are not receiving the error in onyx data, how are you going to access it?

**Comment by alitoshmatov** [comment]
Are you sure, `currentSearchResults.error` contains error that backend sends inside onyxUpdate, right now there is no data inside onyx

<img width="801" alt="Screenshot 2024-12-04 at 10 17 02 PM" src="https://github.com/user-attachments/assets/5a0d4117-1029-48b9-ae3f-f4ff2f4640bb">

**Comment by huult** [comment]
@alitoshmatov I think my alternative solution is suitable for this case. If the user inputs a wrong query and is unable to retrieve any data from onyxData, we should display an empty search view.

**Comment by alitoshmatov** [comment]
Okay, I think we can go with @daledah's [proposal](https://github.com/Expensify/App/issues/52907#issuecomment-2491409822) which suggest these:
- Show empty page when request returns error
- Update `failureData` with custom error to show to user

P. S. for showing error part, we do not have any similar case in search page so this is new thing, we probably need to confirm its design with others. Or another option is just not show error. We just show empty page when request returns error

C+ reviewed 🎀 👀 🎀

**Comment by Julesssss** [comment]
Hey @luacmartins. Sorry to add to your work but we're making search-related changes when request error, and thought you should be notified. 

Do you see any issues with the following for now?:
- Show empty page when request returns error: `Node passed as search filter is invalid for the reportActions table`

**Comment by luacmartins** [comment]
I think that's fine, but maybe we want to display a `no results found` page. cc @Expensify/design for ideas

**Comment by Christinadobrzyn** [comment]
@Julesssss, @luacmartins, @alitoshmatov 

Just to understand, we're thinking this is the flow?  

- Go to staging.new.expensify.com
- Click search router icon.
- Enter the following query: type:chat in: category:car
- Hit Enter.
- Error returned `no results found` - We'd like @Expensify/design to provide feedback on this error.

**Comment by shawnborton** [comment]
Yeah, I think based on what I am seeing above, I would expect us to just show a "Nothing to show" page, like this:
![CleanShot 2024-12-09 at 15 11 39@2x](https://github.com/user-attachments/assets/db66a82e-43de-44b8-b06b-4d839fd91f32)

**Comment by dannymcclain** [comment]
Yeah I agree. I might advocate for updating that copy to something like "Try adjusting your search criteria or creating something with the green + button." for all of these nothing found screens though.

**Comment by Christinadobrzyn** [comment]
Amazing! Thanks @shawnborton and @dannymcclain! Jules does this sound good to move forward with https://github.com/Expensify/App/issues/52907#issuecomment-2524248887

**Comment by dubielzyk-expensify** [comment]
Love the mock and great suggestion Danny 👍

**Comment by Julesssss** [comment]
@Christinadobrzyn yeah, this sounds good to me. Assigning now.

**Comment by alitoshmatov** [comment]
> Yeah I agree. I might advocate for updating that copy to something like "Try adjusting your search criteria or creating something with the green + button." for all of these nothing found screens though.

Several people 👍  this so I assume we are also implementing it.

@daledah Please update `Try creating something with the green + button` to `Try adjusting your search criteria or creating something with the green + button.`

**Comment by Christinadobrzyn** [comment]
Just a heads up that I'm going to be ooo Dec 12 - 13th. Back on Monday 16th. I'm not going to assign this to a BZ teammate but if anything is urgent, please reach out to the team for a volunteer.

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by alitoshmatov** [comment]
> Issue not reproducible during KI retests. (First week)

Issue is still reproducable with `type:chat status:unread in:Mountain category:car` query

**Comment by Christinadobrzyn** [comment]
hi @alitoshmatov and @daledah just checking on the PR for this. Can you provide an update?

**Comment by alitoshmatov** [comment]
PR in progress, waiting @daledah to apply changes to text copies based on slack conversation [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1734059203274889)

**Comment by Christinadobrzyn** [comment]
Payment day 

Contributor: @daledah paid $250 via [Upwork](https://www.upwork.com/nx/wm/offer/105274097)
Contributor+: @alitoshmatov owed $250 via NewDot

@alitoshmatov do we need a regression test for this?

**Comment by alitoshmatov** [comment]
# BugZero Checklist:

- [ ] **[Contributor]** Classify the bug:

<details>
<summary>Bug classification</summary>


Source of bug:
  - [ ] 1a. Result of the original design (eg. a case wasn't considered)
  - [ ] 1b. Mistake during implementation
  - [ ] 1c. Backend bug
  - [ ] 1z. Other:

Where bug was reported:
  - [ ] 2a. Reported on production (eg. bug slipped through the normal regression and PR testing process on staging)
  - [ ] 2b. Reported on staging (eg. found during regression or PR testing)
  - [ ] 2d. Reported on a PR
  - [ ] 2z. Other:

Who reported the bug:
  - [ ] 3a. Expensify user
  - [ ] 3b. Expensify employee
  - [ ] 3c. Contributor
  - [ ] 3d. QA
  - [ ] 3z. Other:

</details>

- [ ] **[Contributor]** The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake.

    Link to comment:

- [ ] **[Contributor]** If the regression was CRITICAL (e.g. interrupts a core flow) A discussion in [#expensify-open-source](https://app.slack.com/client/E047TPA624F/C01GTK53T8Q) has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner.

    Link to discussion:

- [ ] **[Contributor]** If it was decided to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps using the template below to ensure the same bug will not reach production again.

<details>
<summary>Regression Test Proposal Template</summary>


- [ ] **[BugZero Assignee]** Create a GH issue for creating/updating the regression test once above steps have been agreed upon.

    Link to issue:

## Regression Test Proposal
### Precondition:


-

### Test:


1.

Do we agree 👍 or 👎


</details>

**Comment by alitoshmatov** [comment]
@Christinadobrzyn Completed the checklist, requested payment in ND

**Comment by Christinadobrzyn** [comment]
Thanks @alitoshmatov! Regression test [here](https://github.com/Expensify/App/issues/52907#issuecomment-2573602249) 

Payment summary here - https://github.com/Expensify/App/issues/52907#issuecomment-2569461022

Please make sure to reach out in ND for payment @alitoshmatov!

**Comment by JmillsExpensify** [comment]
$250 approved for @alitoshmatov
